<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>phylotypr • phylotypr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="phylotypr">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phylotypr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/phylotypr.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/riffomonas/phylotypr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>phylotypr</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/riffomonas/phylotypr/blob/main/vignettes/phylotypr.Rmd" class="external-link"><code>vignettes/phylotypr.Rmd</code></a></small>
      <div class="d-none name"><code>phylotypr.Rmd</code></div>
    </div>

    
    
<p>In the microbiome field, 16S rRNA gene sequencing has been a popular
method of characterizing the types of bacteria found in a community. The
most frequently used tool for classifying these sequences has been the
<a href="https://journals.asm.org/doi/10.1128/aem.00062-07" class="external-link">naive
Bayesian classifier</a> that used to be accessible through the Ribosomal
Database Project and is currently found in tools like mothur and QIIME.
For the first time, the <a href="https://github.com/riffomonas/phylotypr" class="external-link">phylotypr</a> package provides an
R-based implementation of that tool. There are minimal dependencies and
the easy to use code executes rapidly.</p>
<div class="section level2">
<h2 id="getting-started-with-phylotypr">Getting started with <code>{phylotypr}</code><a class="anchor" aria-label="anchor" href="#getting-started-with-phylotypr"></a>
</h2>
<p>To get started you’ll need to install <a href="https://github.com/riffomonas/phylotypr" class="external-link">phylotypr</a> using
either of these commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#devtools::install_github("riffomonas/phylotypr") #install development version</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"phylotypr"</span><span class="op">)</span> <span class="co">#install stable version from CRAN</span></span></code></pre></div>
<p>To get access to the packages function and a demonstration version of
the training data, you’ll need to run the <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code>
function. Because the classification algorithm makes use of a random
number generator, I strongly encourage you to set the seed to get
consistent results:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/riffomonas/phylotypr" class="external-link">phylotypr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">19760620</span><span class="op">)</span> <span class="co"># pat's birtday in YYYYMMDD format</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="building-the-database">Building the database<a class="anchor" aria-label="anchor" href="#building-the-database"></a>
</h2>
<p>Before we can classify a sequence, we need to generate the database.
This can be done using the <code><a href="../reference/build_kmer_database.html">build_kmer_database()</a></code> function
and takes several seconds to execute. Unless you save the database
object (e.g. <code>db</code> below) to a <code>Rda</code>-formatted
file, you’ll need to run this function each time you load
<a href="https://github.com/riffomonas/phylotypr" class="external-link">phylotypr</a></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_kmer_database.html">build_kmer_database</a></span><span class="op">(</span><span class="va">trainset9_pds</span><span class="op">$</span><span class="va">sequence</span>,</span>
<span>                          <span class="va">trainset9_pds</span><span class="op">$</span><span class="va">taxonomy</span><span class="op">)</span></span></code></pre></div>
<p>By default, <code><a href="../reference/build_kmer_database.html">build_kmer_database()</a></code> will use 8-nt long
kmers to calculate its probabilities. This size can be altered by the
user if needed, but 8-nt performs very well and shouldn’t need to be
adjusted.</p>
</div>
<div class="section level2">
<h2 id="classifying-an-unknown-sequence">Classifying an unknown sequence<a class="anchor" aria-label="anchor" href="#classifying-an-unknown-sequence"></a>
</h2>
<p>To classify a sequence we need the DNA sequence to be a character
object like <code>unknown</code> is below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unknown</span> <span class="op">&lt;-</span> <span class="st">"TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGATCGTTAAGTCAGTGGTCAAATTGAGGGGCTCAACCCCTTCCCGCCATTGAAACTGGCGATCTTGAGTGGAAGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCATGCCGGCTTCCTACTGACGCTCATGCACGAAAGTGTGGGTAACGAACAGG"</span></span></code></pre></div>
<p>The <code>unknown</code> object can then be classified using
<code><a href="../reference/classify_sequence.html">classify_sequence()</a></code> along with the database object from
above (i.e., <code>db</code>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify_sequence.html">classify_sequence</a></span><span class="op">(</span>unknown <span class="op">=</span> <span class="va">unknown</span>, database <span class="op">=</span> <span class="va">db</span><span class="op">)</span></span></code></pre></div>
<p>The output of <code>classify_sequence</code> is a list with taxonomic
levels from kingdom (or domain) down to the genus and the confidence
scores for that sequence in each of those taxonomic levels</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus</span></span>
<span><span class="co">#&gt; $taxonomy</span></span>
<span><span class="co">#&gt; [1] "Bacteria"               "\"Bacteroidetes\""      "\"Bacteroidia\""       </span></span>
<span><span class="co">#&gt; [4] "\"Bacteroidales\""      "\"Porphyromonadaceae\"" "Barnesiella"           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $confidence</span></span>
<span><span class="co">#&gt; [1] 1.00 0.95 0.83 0.83 0.79 0.56</span></span></code></pre></div>
<p>In this example, you’ll notice that the family and genus assignment
have confidence scores below 80%. This is the value commonly used in
mothur and which was used on the RDP website. To filter the
classification to that level of confidence, you can use the
<code><a href="../reference/filter_taxonomy.html">filter_taxonomy()</a></code> function:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_taxonomy.html">filter_taxonomy</a></span><span class="op">(</span><span class="va">consensus</span><span class="op">)</span></span></code></pre></div>
<p>You should see that the “Porphyromonadaceae” and “Barnesiella” names
are removed along with its confidence score.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered</span></span>
<span><span class="co">#&gt; $taxonomy</span></span>
<span><span class="co">#&gt; [1] "Bacteria"          "\"Bacteroidetes\"" "\"Bacteroidia\""  </span></span>
<span><span class="co">#&gt; [4] "\"Bacteroidales\""</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $confidence</span></span>
<span><span class="co">#&gt; [1] 1.00 0.95 0.83 0.83</span></span></code></pre></div>
<p>To convert that list format to a string, you can use the
<code><a href="../reference/print_taxonomy.html">print_taxonomy()</a></code> function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/print_taxonomy.html">print_taxonomy</a></span><span class="op">(</span><span class="va">filtered</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Bacteria(100);\"Bacteroidetes\"(95);\"Bacteroidia\"(83);\"Bacteroidales\"(83);\"Bacteroidales\"_unclassified(83);\"Bacteroidales\"_unclassified(83)"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="more-advanced-topics">More advanced topics<a class="anchor" aria-label="anchor" href="#more-advanced-topics"></a>
</h2>
<div class="section level3">
<h3 id="classifying-multiple-sequences-at-once">Classifying multiple sequences at once<a class="anchor" aria-label="anchor" href="#classifying-multiple-sequences-at-once"></a>
</h3>
<p>If you have multiple sequences that you would like to classify, you
can use the <code><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_chr()</a></code> function from <a href="https://purrr.tidyverse.org/" class="external-link">purrr</a>,
or something like it. Below I show how to classify the same sequence
three times.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unknowns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGATCGTTAAGTCAGTGGTCAAATTGAGGGGCTCAACCCCTTCCCGCCATTGAAACTGGCGATCTTGAGTGGAAGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCATGCCGGCTTCCTACTGACGCTCATGCACGAAAGTGTGGGTAACGAACAGG"</span>,</span>
<span>             <span class="st">"TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGATCGTTAAGTCAGTGGTCAAATTGAGGGGCTCAACCCCTTCCCGCCATTGAAACTGGCGATCTTGAGTGGAAGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCATGCCGGCTTCCTACTGACGCTCATGCACGAAAGTGTGGGTAACGAACAGG"</span>,</span>
<span>             <span class="st">"TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGATCGTTAAGTCAGTGGTCAAATTGAGGGGCTCAACCCCTTCCCGCCATTGAAACTGGCGATCTTGAGTGGAAGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCATGCCGGCTTCCTACTGACGCTCATGCACGAAAGTGTGGGTAACGAACAGG"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_chr</a></span><span class="op">(</span><span class="va">unknowns</span>,</span>
<span>      <span class="op">~</span><span class="fu"><a href="../reference/classify_sequence.html">classify_sequence</a></span><span class="op">(</span>unknown <span class="op">=</span> <span class="va">.x</span>, database <span class="op">=</span> <span class="va">db</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="../reference/filter_taxonomy.html">filter_taxonomy</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="../reference/print_taxonomy.html">print_taxonomy</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Bacteria(100);\"Bacteroidetes\"(100);\"Bacteroidia\"(87);\"Bacteroidales\"(87);\"Bacteroidales\"_unclassified(87);\"Bacteroidales\"_unclassified(87)"</span></span>
<span><span class="co">#&gt; [2] "Bacteria(100);\"Bacteroidetes\"(99);\"Bacteroidia\"(86);\"Bacteroidales\"(86);\"Porphyromonadaceae\"(84);\"Porphyromonadaceae\"_unclassified(84)"    </span></span>
<span><span class="co">#&gt; [3] "Bacteria(100);\"Bacteroidetes\"(100);\"Bacteroidia\"(89);\"Bacteroidales\"(89);\"Porphyromonadaceae\"(84);\"Porphyromonadaceae\"_unclassified(84)"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-to-run-variation">Run-to-run variation<a class="anchor" aria-label="anchor" href="#run-to-run-variation"></a>
</h3>
<p>You’ll notice that the classification of the three sequences varies
slightly. Again, this is due to the use of the random number generator
and getting confidence scores that bounce around 80%, which results in
different levels of filtering. To get a more stable classification, you
should be setting your random number generator seed with
<code><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed()</a></code>. To get a more precise confidence score, you
could set <code>num_bootstraps = 1000</code> in
<code>classify_seqeuence()</code>. Be forewarned that the function will
take 10-times longer to run with 10-times the number of bootstraps!</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_chr</a></span><span class="op">(</span><span class="va">unknowns</span>,</span>
<span>      <span class="op">~</span><span class="fu"><a href="../reference/classify_sequence.html">classify_sequence</a></span><span class="op">(</span>unknown <span class="op">=</span> <span class="va">.x</span>, database <span class="op">=</span> <span class="va">db</span>, num_bootstraps <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="../reference/filter_taxonomy.html">filter_taxonomy</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="../reference/print_taxonomy.html">print_taxonomy</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Bacteria(100);\"Bacteroidetes\"(97.1);\"Bacteroidia\"(87.6);\"Bacteroidales\"(87.6);\"Porphyromonadaceae\"(84.7);\"Porphyromonadaceae\"_unclassified(84.7)"</span></span>
<span><span class="co">#&gt; [2] "Bacteria(100);\"Bacteroidetes\"(96.9);\"Bacteroidia\"(86.1);\"Bacteroidales\"(86.1);\"Porphyromonadaceae\"(83);\"Porphyromonadaceae\"_unclassified(83)"    </span></span>
<span><span class="co">#&gt; [3] "Bacteria(100);\"Bacteroidetes\"(97);\"Bacteroidia\"(84.4);\"Bacteroidales\"(84.4);\"Porphyromonadaceae\"(80.6);\"Porphyromonadaceae\"_unclassified(80.6)"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="alternative-databases">Alternative databases<a class="anchor" aria-label="anchor" href="#alternative-databases"></a>
</h3>
<p>The <a href="https://github.com/riffomonas/phylotypr" class="external-link">phylotypr</a> package ships with the RDP’s v.9 of
their training data. This is relatively small and old (2010) relative to
their latest versions. Also, there are databases available from
greengenes and SILVA. You are encouraged to install newer databases from
the packages on GitHub:</p>
<ul>
<li><a href="https://github.com/mothur/trainset19" class="external-link">RDP’s
<code>{trainset19}</code></a></li>
<li>More on the way!</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pat Schloss.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
